CL2-6708 Avoid inconsistent data check failures related to in_budgeting_participation_context validation

jinjagit opened the pull request:
## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
<br>
No changelog entry included, since this change aims to reduce our reporting of inconsistent data, and is thus of little direct benefit to users and difficult to explain in language that is understandable outside of our dev team.
</details>

- [x] Tests
<details>
<summary>More info</summary>
Added a single test to `spec/models/basket_spec.rb` that confirms a basket is still valid after the project it is associated with is updated to have a non-budgeting `participation_method`, e,g, 'ideation'.

</details>

- [x] Prepared branch for code review


## Links

- [Jira ticket](https://citizenlab.atlassian.net/browse/CL2-6708)

## What changes are in this PR?

We decided to remove the the `in_budgeting_participation_context` validation from the Basket model.

This should avoid the related inconsistent data check failures seen in CI, as in these examples:
```
"issues": {
    ...
    "Basket_participation_context_is_not_budgeting": {
      "count": 27,
      "hosts": {
        "denkmee.t-diel.nl": [
          "11f42fce-a49b-4502-aec4-45b7ff5b7bc1",
          "126c060b-ec46-47bc-8db6-dea24031b7bf",
          "1293cba2-0692-4177-b679-bbbaf71f0e86",
          "1655ff38-ded6-4c53-89ac-2c5a8b2d5a5f",
          "26400a29-5e61-4908-a70e-8cadcbfd10a6",
          "2898dead-b59b-4319-b1f5-0011666b8992",
          "28ae4660-4665-4245-8e09-59e7d5e2d43a",
          "2e2907c0-d1de-4d0f-876c-5e805c9ae101",
          "34bb30a6-748d-45a7-8e55-e2fe34b2159c",
          "42340fc6-c13e-4f23-91b1-4752ca82681c",
          "651a6e2d-1904-4874-af63-2f3f43b23869",
          "66cacf1c-eca0-462d-a768-6e6e7bd5a5f2",
          "82376c41-6796-46b9-baab-8bf465b7e5b0",
          "82908803-4dbc-4298-9010-9539f9f9c2b5",
          "8551b726-eb51-4c18-a67c-29f509c679ee",
          "87250fb9-e195-41b6-ab0c-5f013933d03f",
          "8755ae4b-54fc-4409-8d0e-d1af0a32ed9f",
          "96c41cd5-bb8c-47f3-82e7-7f45f32e5f99",
          "97bbdf63-5427-4fdb-ac15-f1028a2aff53",
          "a436ad59-2468-4f37-980a-3dc85762ba68",
          "a77cd06a-b7e5-48fb-ae00-a579be41c017",
          "cdfe9bb8-b43c-4c76-8780-25ad2d921838",
          "cf780ed6-35c9-4492-ba0c-b47a2154bf14",
          "e3b1fa44-80e4-441f-a653-d83929b67d68",
          "f5220e5e-5a3d-4084-a10f-bca7aba017c4",
          "fef3756d-6f24-4937-9a7e-86d8da7e1f91"
        ],
        "dialogportalen.odense.dk": [
          "c0935be6-2cf1-4bf1-bdd6-8c25d41ddd36"
        ]
      }
```
Excerpt from [this CI run](https://app.circleci.com/pipelines/github/CitizenLabDotCo/citizenlab-ee/4935/workflows/8cc6dce1-1bc8-4506-9cde-a349aed9fbc4/jobs/16876).

## How urgent is a code review?

Medium priority. By the end-of-the week would be nice.

jinjagit committed:
Comment out in_budgeting_participation_context validation

Using CI test runs to check for relevant failures

jinjagit committed:
Add test: failing as desired with validation enabled

Create basket, then upate associated project participation_method
to non-budgeting 'ideation'. Correctly fails with error:
Participation context is not a in budgeting method

jinjagit committed:
Remove in_budgeting_participation_context validation from Basket

Test now passes, as expected.
Also remove debug commands and unused commented out code.

jinjagit committed:
Correct typo in comment

jinjagit commented on back/spec/models/basket_spec.rb:
```diff
@@ -68,4 +68,18 @@
+    # Check the basket remains valid and thus won't fail data consistency checks, as would be the case,
+    # for example, if we enforce validation that the participation_context is budgeting.
+    it "the basket remains valid" do
+      basket = create(:basket, ideas: [idea], participation_context: project, submitted_at: Time.now)
```
I created the basket inside the `it` block, as when the `:in_budgeting_participation_context` is enabled (not removed) this errors nicely, with:
```
1) Basket when the basket's project is updated to non-budgeting participation method the basket remains valid
     Failure/Error: expect(basket).to be_valid
       expected #<Basket id: "af44d31e-a0b2-4f5a-a28d-71ce98e60c39", submitted_at: "2021-08-30 17:34:21.797314000 +0000", user_id: "449efe15-5195-437e-bd1d-7c8b5cd69613", participation_context_id: "ebdc5407-72a1-4a4f-99be-3151d17dc27f", participation_context_type: "Project", created_at: "2021-08-30 17:34:21.970694000 +0000", updated_at: "2021-08-30 17:34:21.970694000 +0000"> to be valid, but got errors: Participation context is not a in budgeting method
```
Whereas, when I declare the basket with `let(:basket) { create(:basket, ideas: [idea], participation_context: project, submitted_at: Time.now) }` _before_ the `it` block, the error occurs before the `expect` assertion:
```
Failures:

  1) Basket when the basket's project is updated to non-budgeting participation method the basket remains valid
     Failure/Error: let(:basket) { create(:basket, ideas: [idea], participation_context: project, submitted_at: Time.now) }
     
     ActiveRecord::RecordInvalid:
       Validation failed: Participation context is not a in budgeting method
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/evaluation.rb:18:in `create'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:12:in `block in result'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:9:in `tap'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:9:in `result'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory.rb:43:in `run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory_runner.rb:29:in `block in run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory_runner.rb:28:in `run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy_syntax_method_registrar.rb:28:in `block in define_singular_strategy_method'
     # ./spec/models/basket_spec.rb:73:in `block (3 levels) in <top (required)>'
     # ./spec/models/basket_spec.rb:82:in `block (3 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.12.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Top 1 slowest examples (0.65629 seconds, 22.5% of total time):
  Basket when the basket's project is updated to non-budgeting participation method the basket remains valid
    0.65629 seconds ./spec/models/basket_spec.rb:79
```

sebastienhoorens replied:
Acceptable to me. If I understand correctly, the issue you're preventing is that you don't want the validation errors to be thrown when doing `create(:basket)`? In that case, a trick that typically works is to not persist the instance yet in the database. You can do this with `build(:basket)`. Then validations will only run for the first time when doing `basked.valid?`, which is I think what you want. The disadvantage however, is that you wont be able to create proper relationships with other instances (as this requires the basket to be persisted in the database).

sebastienhoorens replied:
Ah no, I understand the issue now. The problem is that `let` is lazy in evaluation. This means the basket will only be created when the line `basket.reload` is evaluated, which is too late and therefore throws the error. You can use `let!` instead and you will see that it then works (the basket will be created before the project update). But it's still fine to also create the basket within the `it` block :)

jinjagit replied:
@sebastienhoorens Ah, that's a great explanation. :+1: Thanks. I'll think about which I prefer and merge when I decide (soon).

jinjagit committed:
Use immediate evaluation of basket declaration, not lazy

This gives a cleaner error message if the basket fails validation,
which is what this test checks for.
